### PR TITLE
Ignore script fields when size is 0

### PR DESCRIPTION
**Pull request by Parth-Vader**
Addresses issue #31824 

I have added the condition as discussed.

### CODE REVIEW

**Review by cbuescher — CHANGES_REQUESTED** [review]
Hi @Parth-Vader, thanks for opening the PR. Could you also add a unit test for the change you made? I think adding a test case to SearchServiceTests would work. You can take a look at e.g the existing "testMaxScriptFieldsSearch()" which tests something slightly different but covers the same area in the code.

**Comment by Parth-Vader** [comment]
@cbuescher I tried adding a unit test, but I have one problem. 
What is the default return of `context.scriptFields()` when the source script size is 0, and it is skipped?
Thanks.

**Comment by Parth-Vader** [comment]
@cbuescher @colings86 Can you have a look at the test? Is it correctly done? Thanks.

File: server/src/test/java/org/elasticsearch/search/SearchServiceTests.java [hunk 1]
```diff
@@ -345,6 +346,21 @@ public void testMaxScriptFieldsSearch() throws IOException {
...
+    public void testIgnoreScriptfieldIfSizeZero() throws IOException{
+        createIndex("index");
+        final SearchService service = getInstanceFromNode(SearchService.class);
+        final IndicesService indicesService = getInstanceFromNode(IndicesService.class);
+        final IndexService indexService = indicesService.indexServiceSafe(resolveIndex("index"));
+        final IndexShard indexShard = indexService.getShard(0);
+
+        SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder();
+        searchSourceBuilder.size(0);
+        try(SearchContext context = service.createContext(new ShardSearchLocalRequest(indexShard.shardId(), 1, SearchType.DEFAULT,
```
Comment (cbuescher) [inline server/src/test/java/org/elasticsearch/search/SearchServiceTests.java]: nit: whitespace after "try"

File: server/src/test/java/org/elasticsearch/search/SearchServiceTests.java [hunk 2]
```diff
@@ -345,6 +346,21 @@ public void testMaxScriptFieldsSearch() throws IOException {
...
+                searchSourceBuilder, new String[0], false, new AliasFilter(null, Strings.EMPTY_ARRAY), 1.0f, true, null, null))) {
+                    assertEquals(0,context.scriptFields().fields().size());
```
Comment (cbuescher) [inline server/src/test/java/org/elasticsearch/search/SearchServiceTests.java]: nit: whitespace after comma

**Review by cbuescher — CHANGES_REQUESTED** [review]
Hi @Parth-Vader, the test already looks quite good in general. However, it currently also passes without your changes in SearchService.java. I think you need to add a script field to searchSourceBuilder, and later you then check that this is ignored when size=0. You can take a look at "testMaxScriptFieldsSearch" (L328) for an example of how to add a dummy script field to the search source.

**Comment by Parth-Vader** [comment]
@cbuescher Hi, I added a script field, then set the size to zero. Also added the required white spaces.

File: server/src/test/java/org/elasticsearch/search/SearchServiceTests.java [hunk 3]
```diff
@@ -345,6 +346,26 @@ public void testMaxScriptFieldsSearch() throws IOException {
...
 
+    public void testIgnoreScriptfieldIfSizeZero() throws IOException{
+        createIndex("index");
+        final SearchService service = getInstanceFromNode(SearchService.class);
+        final IndicesService indicesService = getInstanceFromNode(IndicesService.class);
+        final IndexService indexService = indicesService.indexServiceSafe(resolveIndex("index"));
+        final IndexShard indexShard = indexService.getShard(0);
+
+        SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder();
+
```
Comment (cbuescher) [inline server/src/test/java/org/elasticsearch/search/SearchServiceTests.java]: nit: pls remove the extra blank lines (at least the double one)

File: server/src/test/java/org/elasticsearch/search/SearchServiceTests.java [hunk 4]
```diff
@@ -345,6 +346,26 @@ public void testMaxScriptFieldsSearch() throws IOException {
...
+        SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder();
+        
+        
+        searchSourceBuilder.scriptField("field" + 0,
+                new Script(ScriptType.INLINE, MockScriptEngine.NAME, CustomScriptPlugin.DUMMY_SCRIPT, Collections.emptyMap()));
+        
+        searchSourceBuilder.size(0);
+        try (SearchContext context = service.createContext(new ShardSearchLocalRequest(indexShard.shardId(), 1, SearchType.DEFAULT,
+                searchSourceBuilder, new String[0], false, new AliasFilter(null, Strings.EMPTY_ARRAY), 1.0f, true, null, null))) {
+                    assertEquals(0, context.scriptFields().fields().size());
```
Comment (cbuescher) [inline server/src/test/java/org/elasticsearch/search/SearchServiceTests.java]: nit: please move indentation to the right so it alligns at four spaces in the try-block)

**Review by cbuescher — CHANGES_REQUESTED** [review]
Hi @Parth-Vader, thanks for the changes, there are only a couple of very minor formatting changes I'd like you to make if possible. The rest looks great now.

File: server/src/test/java/org/elasticsearch/search/SearchServiceTests.java (hunk 4 above)
Comment (Parth-Vader) [reply server/src/test/java/org/elasticsearch/search/SearchServiceTests.java]: sorry, do you mean shift `assertEquals` four spaces backwards?

**Review by cbuescher — APPROVED** [review]
@Parth-Vader many thanks, LGTM. I will run CI before merging.

**Comment by cbuescher** [comment]
@elasticmachine test this please

File: server/src/test/java/org/elasticsearch/search/SearchServiceTests.java [hunk 5]
```diff
@@ -56,6 +56,7 @@
 import org.elasticsearch.search.aggregations.support.ValueType;
 import org.elasticsearch.search.builder.SearchSourceBuilder;
 import org.elasticsearch.search.fetch.ShardFetchRequest;
+import org.elasticsearch.search.fetch.subphase.ScriptFieldsContext;
```
Comment (cbuescher) [inline server/src/test/java/org/elasticsearch/search/SearchServiceTests.java]: Looks like the checkstyle runner doesn't like unused imports, could you remove this line please?

```
17:52:18 [ant:checkstyle] [ERROR] /var/lib/jenkins/workspace/elastic+elasticsearch+pull-request/server/src/test/java/org/elasticsearch/search/SearchServiceTests.java:59:8: Unused import - org.elasticsearch.search.fetch.subphase.ScriptFieldsContext. [UnusedImports]
```

Comment (Parth-Vader) [reply server/src/test/java/org/elasticsearch/search/SearchServiceTests.java]: My bad. Done!

**Comment by cbuescher** [comment]
@elasticmachine test this please

**Comment by cbuescher** [comment]
@Parth-Vader more CI failures, this time due to an old version of master. I took the liberty to merge in current master, hope you don't mind
@elasticmachine test this please

**Comment by cbuescher** [comment]
@Parth-Vader thanks again, I merged this to master and the 6.5 branch.

**Comment by Parth-Vader** [comment]
@cbuescher Thanks for guiding me through this.

**Pull request merged**